### ISSUE TITLE
UKSI update: Duplicate ladybird name

### ISSUE THREAD

**Issue by robin-hutchinson** [BiologicalRecordsCentre/iRecord#1476]
Nephus redtenbacheri appears in the UKSI twice, but with different TVKs and organism keys as one of the names is a junior synonym of Nephus limonii/redtenbacheri agg.: https://uksi-sandbox.nhm.ac.uk/taxonbrowser.php?txtSearch=Nephus+redtenbacheri

Both of these names appear in the drop down menu on recording forms:

![image](https://github.com/BiologicalRecordsCentre/iRecord/assets/95474755/3aabec73-c2c5-4cce-adb0-afc27d76f520)

Checking Indicia, both are tagged as preferred names. The taxon with TVK "NBNSYS0000008306"  and recommended TVK "NHMSYS0021366799" should be marked as preferred = false and not for data entry.

**Comment by robin-hutchinson** [comment]
Just an update on this, it looks like there are 321 instances where allow_data_entry = TRUE and preferred = TRUE, but the name_tvk does not match the rtvk, so something may have happened here?

select cttl.kingdom_taxon, cttl.order_taxon, cttl.family_taxon, cttl.taxon, cttl.authority, cttl.taxon_group, cttl.taxon_rank, cttl.taxon_rank_sort_order, cttl.preferred_taxon, cttl.preferred_authority, cttl.preferred, cttl.default_common_name common_name, cttl.search_code as name_tvk, cttl.external_key as rtvk, t.organism_key
from cache_taxa_taxon_lists cttl
join taxa t 
on cttl.taxon_id = t.id
where taxon_list_id=15
and cttl.language = 'Latin'
and cttl.allow_data_entry = TRUE
and cttl.preferred = TRUE
and cttl.search_code != cttl.external_key
order by kingdom_taxon, order_taxon, family_taxon, preferred_taxon, preferred desc

**Comment by johnvanbreda** [comment]
@burkmarr can I ask you to check my thinking? In script 23, we prepare a list of proposed taxa taxon list data for comparison with the actual live data, so that we can later insert or update the live data to match. This sets the preferred flag to true only for records where the search code (name TVK) matches the external key (accepted or rTVK). It also creates a null field called orig_preferred which gets updated with the data from the live table in script 24. At this point, a flag called changed gets set to true if the allow data entry value is changing, but it doesn't get set to true the preferred flag is changing and I think it should. Later, the changed flag gets used to filter which records actually need to be updated, so if a taxa taxon list record is identical other than the preferred flag changing it won't get updated. 

If I'm right, it means a small change to script 24 and I think we should be able to write a query that fixes the data without re-running everything.

**Comment by burkmarr** [comment]
@johnvanbreda - I've tried to get my head around this. Your suggestion to set the changed flag to true if the preferred flag is changing seems right to me. But I can't see how the current scripts could have led to instances where the preferred flag is set to true when the search code does not match the external key. (Apart from the 321 identified by Robin there are a further 64 where allow_data_entry is false.)

**Comment by johnvanbreda** [comment]
Aren't they either cases where they used to be preferred, but the preferred name has changed, so the external key changed (rTVK) but the preferred flag was left as it was for the former accepted name that was now relegated?

**Comment by burkmarr** [comment]
@johnvanbreda - yes I think I understand now.

**Comment by burkmarr** [comment]
@johnvanbreda 

> I think we should be able to write a query that fixes the data without re-running everything

Is this just a simple case of resetting the preferred and allow data entry flags to false for those taxa identified above?

**Comment by johnvanbreda** [comment]
Yes. We should probably also update any linked occurrences using the work queue as well as update the taxa.yml lookup file used by Logstash. Let me know if you want any help or checks of proposed actions.

**Comment by burkmarr** [comment]
@johnvanbreda - can you check the steps below?

Select affected ttl id's into a temporary table:
```sql
select id into iss_1476_affected_ttl from cache_taxa_taxon_lists  
where taxon_list_id=15 and preferred = TRUE and search_code != external_key 
```
Update the ttl table:
```sql
update taxa_taxon_lists set preferred = FALSE, updated_on=Now(), updated_by_id=198219 
where id in (select id from iss_1476_affected_ttl) 
```
Update the cttl table:
```sql
update cache_taxa_taxon_lists set preferred = FALSE, updated_on=Now(), updated_by_id=198219 
where id in (select id from iss_1476_affected_ttl) 
```
Create work_queue tasks for affected occurrences (there are 42310 of them):
```sql
insert into work_queue(task, entity, record_id, params, cost_estimate, priority, created_on) 
select 'task_cache_builder_update', 'occurrence', id, null, 100, 2, now() 
from occurrences where taxa_taxon_list_id in (select id from iss_1476_affected_ttl) 
```
Update the ES taxa.yml file (see https://github.com/Indicia-Team/support_files/blob/master/Elasticsearch/docs/occurrences.md), replace on the ES server & restart the logstash service.

**Comment by johnvanbreda** [comment]
That all looks fine to me, except I'd update the taxa.yml file and restart logstash before inserting the log queue entries, so that there is no chance of a queue entry being processed before the taxonomy is updated. A couple of minor suggestions would be to use a temporary table instead of a permanent one in the first step and to use the taxonomy raw tables instead of the cache copies so we are looking at the data in it's most precise form:

```sql
select ttl.id 
into temporary iss_1476_affected_ttl 
from taxa_taxon_lists ttl
join taxa t on t.id=ttl.taxon_id and t.deleted=false
where ttl.taxon_list_id=15 and ttl.preferred = TRUE and t.search_code<>t.external_key 
and ttl.deleted=false
```

I had a look at the reverse case, where preferred is false but the 2 keys are the same, but the cases I found were for old deprecated names so not likely to be a real problem. I did find and fix one problem in the UKSI history incremental updates though!

**Comment by burkmarr** [comment]
I've done this update now. There were a couple of mistakes in the middle two SQL queries above - https://github.com/BiologicalRecordsCentre/iRecord/issues/1476#issuecomment-1564278953 - so I corrected the comment (I had forgotten to update the allow_data_entry flag).

I did as @johnvanbreda suggested and selected from raw tables instead of cache table. But I didn't use a temporary table on this occasion because I was worried about it being lost if I was booted out of my remote desktop session which has happened before now. I will delete the table if all is well. (I deleted some tables I had left behind from dealing with previous issues.) I also followed @johnvanbreda suggestion to update the Logstash taxa.yml file (and restart the service) before generated the work_queue cache builder entries.

The data entry problem is fixed now @robin-hutchinson. It will take a while for the affected occurrences to be updated on ES as the scheduled tasks get through the work queue.

**Comment by robin-hutchinson** [comment]
Brilliant, thank you so much, I'll let the recorder know (once the change has taken effect!)

**Comment by burkmarr** [comment]
Changes described above are discussed further in issue 1575 and plan is to reverse them: https://github.com/BiologicalRecordsCentre/iRecord/issues/1575#issuecomment-1761486314
